drivers/gpio/gpio_emul.c

```diff
@@ -0,0 +1,687 @@
+ * model GPIO interrupt controllers with varying interrupt trigger support.
+ *
+ * For example, some controllers to not support level interrupts,
+ * some controllers do not support rising and falling edge simultaneously,
+ * etc.
+ *
+ * This primarily affects the behaviour of @ref gpio_pin_interrupt_configure.
+ */
+enum gpio_emul_interrupt_cap {
+	GPIO_EMUL_INT_CAP_EDGE_RISING = 1,
+	GPIO_EMUL_INT_CAP_EDGE_FALLING = 2,
+	GPIO_EMUL_INT_CAP_EDGE_BOTH = 4,
+	GPIO_EMUL_INT_CAP_LEVEL_HIGH = 16,
+	GPIO_EMUL_INT_CAP_LEVEL_LOW = 32,
+	GPIO_EMUL_INT_CAP_LEVEL_BOTH = 64,
+};
+
+/**
+ * @brief Emulated GPIO controller configuration data
+ *
+ * This structure contains all of the state for a given emulated GPIO
+ * controller as well as all of the pins associated with it.
+ *
+ * The @a flags member is a pointer to an array which is @a num_pins in size.
+ *
+ * @a num_pins must be in the range [1, @ref GPIO_MAX_PINS_PER_PORT].
+ *
+ * Pin direction as well as other pin properties are set using
+ * specific bits in @a flags. For more details, see @ref gpio_interface.
+ *
+ * Changes are synchronized using @ref gpio_emul_data.mu.
+ */
+struct gpio_emul_config {
+	/** Common @ref gpio_driver_config */
+	const struct gpio_driver_config common;
+	/** Number of pins available in the given GPIO controller instance */
+	const gpio_pin_t num_pins;
+	/** Supported interrupts */
+	const enum gpio_emul_interrupt_cap interrupt_caps;
+};
+
+/**
+ * @brief Emulated GPIO controller data
+ *
+ * This structure contains data structures used by a emulated GPIO
+ * controller.
+ *
+ * If the application wishes to specify a "wiring" for the emulated
+ * GPIO, then a @ref gpio_callback_handler_t should be registered using
+ * @ref gpio_add_callback.
+ *
+ * Changes are to @ref gpio_emul_data and @ref gpio_emul_config are
+ * synchronized using @a mu.
+ */
+struct gpio_emul_data {
+	/** Common @ref gpio_driver_data */
+	struct gpio_driver_data common;
+	/** Pointer to an array of flags is @a num_pins in size */
+	gpio_flags_t *flags;
+	/** Input values for each pin */
+	gpio_port_value_t input_vals;
+	/** Output values for each pin */
+	gpio_port_value_t output_vals;
+	/** Interrupt status for each pin */
+	gpio_port_pins_t interrupts;
+	/** Mutex to synchronize accesses to driver data and config */
+	struct k_mutex mu;
+	/** Singly-linked list of callbacks associated with the controller */
+	sys_slist_t callbacks;
+};
+
+/**
+ * @brief Obtain a mask of pins that match all of the provided @p flags
+ *
+ * Use this function to see which pins match the current GPIO configuration.
+ *
+ * The caller must ensure that @ref gpio_emul_data.mu is locked.
+ *
+ * @param port The emulated GPIO device pointer
+ * @param mask A mask of flags to match
+ * @param flags The flags to match
+ *
+ * @return a mask of the pins with matching @p flags
+ */
+static gpio_port_pins_t
+get_pins_with_flags(const struct device *port, gpio_port_pins_t mask,
```

Comment:
`gpio_flags_t` mask, since the input mask is also flags of interest, not pins of interest.

Reply:
Yes - thanks!